solr/core/src/java/org/apache/solr/core/CoreContainer.java

```diff
@@ -168,6 +172,33 @@ public CoreLoadFailure(CoreDescriptor cd, Exception loadFailure) {
     }
   }
 
+  public static class ClusterSingletons {
```

Comment:
There is no need for `CoreContainer` to be aware of `ClusterSingleton`

Only `ContainerPlugins` and `Overseer` need to be aware of it

Reply:
That's the downside of splitting this PR into API & implementation ... The way we load CC it's nearly impossible to properly initialize singletons from within the load() method if we don't defer the initialization until the load() is completed. `ClusterSingletons` class is just a helper for deferred init of singletons once they are all discovered and loaded.

Reply:
>That's the downside of splitting this PR into API & implementation ..

I thouh this Pr has both API and impl

`CC.load()` is a kitchen sink today. We should not need such large changes to CC everytime something is added

Reply:
Moved most of that into a utility class.